src/rbindlist.c

```diff
@@ -310,10 +310,21 @@ SEXP rbindlist(SEXP l, SEXP usenamesArg, SEXP fillArg, SEXP idcolArg)
       if (firsti==-1) { firsti=i; firstw=w; firstCol=thisCol; }
       else {
         if (!factor && !int64) {
-          if (!R_compute_identical(PROTECT(getAttrib(thisCol, R_ClassSymbol)),
-                                   PROTECT(getAttrib(firstCol, R_ClassSymbol)),
-                                   0)) {
-            error(_("Class attribute on column %d of item %d does not match with column %d of item %d."), w+1, i+1, firstw+1, firsti+1);
+          SEXP thisClass = PROTECT(getAttrib(thisCol, R_ClassSymbol));
+          SEXP firstClass = PROTECT(getAttrib(firstCol, R_ClassSymbol));
+          if (!R_compute_identical(thisClass, firstClass, 0)) {
+            #define BUFSIZE 1024 // 'error' function length limit in R's errors.c is 8192
+            char thisClassBuf[BUFSIZE];
+            char* thisClassStr = concatCharVec(thisClass, ", ");
+            strncpy(thisClassBuf, thisClassStr, BUFSIZE);
+            thisClassBuf[BUFSIZE-1] = '\0';
+            free(thisClassStr);
+            char firstClassBuf[BUFSIZE];
+            char* firstClassStr = concatCharVec(firstClass, ", ");
+            strncpy(firstClassBuf, firstClassStr, BUFSIZE);
```

Comment:
`base` `do_paste` uses `strcpy`, is that more appropriate? In particular I think `strcpy` handles the terminal `'\0'` right? I am not an expert on the difference.

Reply:
`strcpy` will attempt to copy the source string in its entirety, even if the destination array is too small.  `strncpy` adds a size argument and will stop copying when that size limit is reached.  If the string terminates before the size limit is reached, then the `'\0'` byte will be copied, otherwise it needs to be added manually.

Since we don't know how large the string returned by `concatCharVec` will be, this seemed like the right thing to do.  An alternative option would be to add a `max_size` argument to `concatCharVec` to limit the size of the output.

Reply:
thanks for the explanation, makes sense. I don't think a new argument is necessary.